### ISSUE TITLE
Updated to MM4 but SSL not working

### ISSUE THREAD

**Issue by beaulidden** [middleman/middleman#2052]
Trying to run a simple `middleman init` which works fine, until I get the following error:

`Could not load OpenSSL. You must recompile Ruby with OpenSSL support or change the sources in your Gemfile from 'https' to 'http'. Instructions for compiling with OpenSSL using RVM are available at rvm.io/packages/openssl.`

I'm aware it may be $PATH related but I don't know how to change that. When I run `which openssl` I get `/usr/bin/openssl`

So what I *think* I need to do is update my openssl version, somehow, someway which will then allow me to perform a successful `middleman init` and `bundle install` afterwards.

Where should I start? Any help would be amazing :) thanks.

**Comment by vvasabi** [comment]
This appears to be an RVM-related issue, and not MM’s. I would start by upgrading RVM to the latest version and reinstalling Ruby afterwards. Newer RVM versions should be smart enough to detect the installation of OpenSSL (or the lack of).

**Comment by beaulidden** [comment]
No luck. I ran `rvm get stable` which, after execution, showed the following via `rvm version`:

`rvm 1.29.0 (latest) by Michal Papis, Piotr Kuczynski, Wayne E. Seguin [https://rvm.io/]`

Then, I ran `rvm install 2.4.0` which ensures I was running the latest version of Ruby.

But I'm still seeing the OpenSSL issue.

**Comment by vvasabi** [comment]
Does `ruby -v` also say that you are indeed using the newly installed version?

**Comment by beaulidden** [comment]
Yep. Check it out:  https://imgur.com/a/wQwTe

**Comment by vvasabi** [comment]
What about `which middleman` and `middleman version`? I’m just throwing out things I would check.

**Comment by beaulidden** [comment]
```
Last login: Tue Feb 14 10:23:37 on ttys000
Beaus-MacBook-Pro:~ beau$ which middleman
/Users/beau/.rvm/rubies/ruby-2.4.0/bin/middleman
Beaus-MacBook-Pro:~ beau$ middleman version
Middleman 4.2.1
Beaus-MacBook-Pro:~ beau$ 
```

**Comment by vvasabi** [comment]
Can you try this command?

```
rvm osx-ssl-certs update all
```

It fixed an OpenSSL issue for me in the past (but with a different error message).

**Comment by beaulidden** [comment]
The command seemed to work, but the error persists. FWIW here's how I'm generating the error, while trying to get middleman into a folder:

```
Beaus-MacBook-Pro:untitled folder beau$ middleman init
         run  git clone --depth 1 https://github.com/middleman/middleman-templates-default.git /var/folders/_z/nr8vt7gs2r1g0hwlsvlnx72m0000gn/T/d20170214-44963-iusqrb from "."
Cloning into '/var/folders/_z/nr8vt7gs2r1g0hwlsvlnx72m0000gn/T/d20170214-44963-iusqrb'...
remote: Counting objects: 22, done.
remote: Compressing objects: 100% (15/15), done.
remote: Total 22 (delta 0), reused 15 (delta 0), pack-reused 0
Unpacking objects: 100% (22/22), done.
       exist  
      create  .gitignore
      create  Gemfile
      create  config.rb
      create  source/images/middleman-logo.svg
      create  source/index.html.erb
      create  source/javascripts/all.js
      create  source/layouts/layout.erb
      create  source/stylesheets/_normalize.scss
      create  source/stylesheets/site.css.scss
Do you want to use Compass? yes
Do you want to use LiveReload? yes
      insert  Gemfile
      insert  Gemfile
      insert  config.rb
      insert  Gemfile
Do you want a Rack-compatible config.ru file? yes
      create  config.ru
         run  bundle install from "."
The dependency wdm (~> 0.1.0) will be unused by any of the platforms Bundler is installing for. Bundler is installing for ruby but the dependency is only for x86-mswin32, x86-mingw32. To add those platforms to the bundle, run `bundle lock --add-platform x86-mswin32 x86-mingw32`.
The dependency tzinfo-data (>= 0) will be unused by any of the platforms Bundler is installing for. Bundler is installing for ruby but the dependency is only for x86-mswin32, x86-mingw32, java. To add those platforms to the bundle, run `bundle lock --add-platform x86-mswin32 x86-mingw32 java`.
Could not load OpenSSL.
You must recompile Ruby with OpenSSL support or change the sources in your
Gemfile from 'https' to 'http'. Instructions for compiling with OpenSSL using
RVM are available at rvm.io/packages/openssl.
Beaus-MacBook-Pro:untitled folder beau$ 
```
I really appreciate all your help on this. Means a lot and is inching me closer to getting MM up and running so I can push a ton of updates to a site I run and manage :)

**Comment by vvasabi** [comment]
I’m out of ideas here. :-/ See if this post can help you: http://railsapps.github.io/openssl-certificate-verify-failed.html

**Comment by beaulidden** [comment]
:'( well I'm officially screwed hah. All good, thanks for all your help anyway :) I tried the link but still no resolve.

**Comment by stevenosloan** [comment]
From what I can tell, you've essentially found the "conical" post for [this specific issue](http://railsapps.github.io/openssl-certificate-verify-failed.html) -- if none of that works for you one options you *could* try would be to try changing the [ruby gem source](https://middleman/blob/master/middleman-cli/lib/middleman-cli/templates/extension/Gemfile#L2) in the Gemfile to `http://` to avoid using openssl. That said I would **not** suggest this, as you lose any protection from man-in-the middle, tracking, etc.

Every time I've had this issue personally, I've reinstalled the specific ruby verison via rvm and that has resolved it, but clearly YMMV.

**Comment by iwarner** [comment]
On mac I usually also 

brew install openssl 
and force link this version

http://stackoverflow.com/questions/15185661/update-openssl-on-os-x-with-homebrew

You can give this a go maybe - what OSX you on?

**Comment by beaulidden** [comment]
Cool, I'll give it a shot. I'm on macOS Sierra(10.12.3)

**Comment by beaulidden** [comment]
P.S. I think we've come across one another before @iwarner...probably a while back now. Thanks for helping me out again :)

**Comment by iwarner** [comment]
Sounds mysterious.
On sierra actually I did not need to do this openssl trick this time, but who knows

I also use Rbenv - just easier to control I think

/Users/ianwarner/.rbenv/shims/ruby
ruby 2.3.1p112 (2016-04-26 revision 54768) [x86_64-darwin16]

/Users/ianwarner/.rbenv/shims/gem
gem 2.6.8

**Comment by beaulidden** [comment]
This whole thing is mysterious. It's doing my head in :(

Odd, my Ruby is way behind yours

```
Beaus-MacBook-Pro:test beau$ ruby -v
ruby 2.0.0p648 (2015-12-16 revision 53162) [universal.x86_64-darwin16]
```

**Comment by iwarner** [comment]
Yeah I always keep mine up to date.

Switch to Rbenv perhaps and install the stable version

    # Variables
    VERSION=2.3.1

    # Header
    header "RBENV Installation Script"

    log "Continue (y/Y)"
    read RBENV_CONTINE

    if [[ $RBENV_CONTINE == y || $RBENV_CONTINE == Y ]] ; then

        subheader "Brew Update"
        brew update

        subheader "Install Rbenv, Rbenv Gem and Ruby Build"
        brew install rbenv ruby-build

        eval "$(rbenv init - --no-rehash)"

        # Install Ruby Version
        subheader "Install Ruby Version"

        CONFIGURE_OPTS="--disable-install-doc --with-readline-dir=$(brew --prefix readline)" rbenv install $VERSION

        rbenv global $VERSION
        rbenv rehash

        # Remove User Gems folder
        rm -r ~/.gem

    fi

**Comment by beaulidden** [comment]
Cool. [Official rbenv documentation](https://github.com/rbenv/rbenv#installing-ruby-versions) is suggesting this as a way to update to a specific version:

`rbenv install <version>`

After running `rbenv install -l` it looks as though 2.4.0 is the latest. All good?

**Comment by iwarner** [comment]
Indeed thats how I install outside the script - the script just helps with the initial added options.

**Comment by beaulidden** [comment]
Well this is confusing. I just did a fresh install of 2.4.0 but then a `ruby -v` returns

`ruby 2.0.0p648 (2015-12-16 revision 53162) [universal.x86_64-darwin16]`

Full output as follows:

```
Beaus-MacBook-Pro:test beau$ rbenv install 2.4.0
ruby-build: use openssl from homebrew
Downloading ruby-2.4.0.tar.bz2...
-> https://cache.ruby-lang.org/pub/ruby/2.4/ruby-2.4.0.tar.bz2
Installing ruby-2.4.0...
ruby-build: use readline from homebrew
Installed ruby-2.4.0 to /Users/beau/.rbenv/versions/2.4.0

Beaus-MacBook-Pro:test beau$ ruby -v
ruby 2.0.0p648 (2015-12-16 revision 53162) [universal.x86_64-darwin16]
```

**Comment by iwarner** [comment]
Yeah the universal means you are using the Ruby provided by the OS

do - echo $PATH and make sure homebrew is in the path and in the correct place.

Oherwise ammend your Path - maybe reinstall Homebrew also

see if this brings up anything
brew doctor

**Comment by beaulidden** [comment]
echo $PATH didn't reveal anything to do with Brew, oddly enough:

```
Beaus-MacBook-Pro:test beau$ echo $PATH
/Users/beau/.rvm/gems/ruby-2.4.0/bin:/Users/beau/.rvm/gems/ruby-2.4.0@global/bin:/Users/beau/.rvm/rubies/ruby-2.4.0/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/Users/beau/.rvm/bin
```
Running `brew doctor` worked and suggested pruning, which is what I did:

```
Beaus-MacBook-Pro:test beau$ brew doctor
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry and just ignore them. Thanks!

Warning: Broken symlinks were found. Remove them with `brew prune`:
  /usr/local/bin/openssl
Beaus-MacBook-Pro:test beau$ brew prune
Pruned 1 symbolic links and 4 directories from /usr/local
```

So it looks as though Brew isn't even in the path...if that is a thing? Do you suggest completely removing and re-installing Brew?

**Comment by iwarner** [comment]
Gotcha - if you are not familiar with Path manipulation then I would point you towards reading up on dotfiles.

It will also open a new world for you in terms of terminal usage - https://dotfiles.github.io/ - you probably dont have ZSH and auto complete etc... ??

Below is my entry in my .zshrc file - after reading above you will know what that means :)

# Homebrew
PATH=/usr/local/sbin:
PATH=$PATH/usr/local/bin:

# Default
PATH=$PATH/usr/bin:
PATH=$PATH/usr/sbin:
PATH=$PATH/bin:
PATH=$PATH/sbin:

# Laravel
PATH=$PATH~/.composer/vendor/bin:

# Android SDK Path - generic for Homebrew install - brew install android-sdk
PATH=$PATH/usr/local/opt/android-sdk

# Export path
export path

**Comment by iwarner** [comment]
On further looking at your path - /usr/local/bin - this is what homebrew needs.

**Comment by beaulidden** [comment]
Cool, I'll read up on the dotfiles. Looks like there's a lot of managers to choose from. Do you recommend one? Any no - I don't have ZSH or auto complete. I'm only a part time front-end developer on the side of my full time job as a UI/product manager. So I'm a bit of a rookie...yet somehow I've been able to roll with Middleman seemingly unscathed for the better of the last 3 or os years. All of a sudden I'm stuck eh!

**Comment by beaulidden** [comment]
P.S. I'm in Australia so will be heading off soon, but will post back with how I go in the morning and when your back in action in London town we can go from there. Thanks again for your help. I really appreciate it :)

**Comment by beaulidden** [comment]
Hi everyone. I ended up solving most of the issues described above by completely removing all my gems, and uninstalling rbenv _and_ rvm which, I'm guessing, was the root of a lot of my problems (as in the two being installed and modified at once)

After using rbenv to update Ruby versions, and switching to that instead of the OS X system Ruby version, I was able to do something that I haven't been able to do in two weeks; *install middleman!*

Thanks a bunch to everyone above who helped me achieve this. Seeing this community step up and help folks like me really means a lot and keeps my faith in the dev community shining bright :) I've come up against another little snag but I'll open a new ticket for that one which, relatively speaking, isn't as curly as the issue here.

Thanks again!
